File: common/src/main/java/juuxel/adorn/item/TradingStationItem.java
package juuxel.adorn.item;

import juuxel.adorn.component.AdornComponentTypes;
import net.minecraft.world.item.Item.Properties;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.item.BlockItem;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.inventory.tooltip.TooltipComponent;

import java.util.Optional;

public final class TradingStationItem extends BlockItem {
    public TradingStationItem(Block block, Properties settings) {
        super(block, settings);
    }

    @Override
    public boolean canFitInsideContainerItems() {
        // Don't allow putting trading stations inside shulker boxes or other trading stations.
        return false;
    }

    @Override
    public Optional<TooltipComponent> getTooltipImage(ItemStack stack) {
        var trade = stack.get(AdornComponentTypes.TRADE.get());
        if (trade != null && !trade.isFullyEmpty()) {
            return Optional.of(trade);
        }

        return Optional.empty();
    }
}

File: common/src/main/kotlin/juuxel/adorn/item/TradingStationItem.kt
package juuxel.adorn.item

import juuxel.adorn.block.entity.TradingStationBlockEntity
import juuxel.adorn.trading.Trade
import juuxel.adorn.util.getCompoundOrNull
import net.minecraft.block.Block
import net.minecraft.client.item.TooltipData
import net.minecraft.item.ItemStack
import java.util.Optional

class TradingStationItem(block: Block, settings: Settings) : BaseBlockItem(block, settings) {
    // Don't allow putting trading stations inside shulker boxes or other trading stations.
    override fun canBeNested(): Boolean = false

    override fun getTooltipData(stack: ItemStack): Optional<TooltipData> {
        stack.getSubNbt(BLOCK_ENTITY_TAG_KEY)?.let { nbt ->
            nbt.getCompoundOrNull(TradingStationBlockEntity.NBT_TRADE)?.let { tradeNbt ->
                val trade = Trade.fromNbt(tradeNbt)
                if (!trade.isFullyEmpty()) {
                    return Optional.of(trade)
                }
            }
        }
        return Optional.empty()
    }
}
